python/gstaichi/lang/_fast_caching/args_hasher.py

```diff
@@ -21,25 +21,28 @@
 FIELD_METADATA_CACHE_VALUE = "add_value_to_cache_key"
 
 
-def dataclass_to_repr(arg: Any) -> str:
+def dataclass_to_repr(path: tuple[str, ...], arg: Any) -> str:
     repr_l = []
     for field in dataclasses.fields(arg):
         child_value = getattr(arg, field.name)
-        _repr = stringify_obj_type(child_value)
+        _repr = stringify_obj_type(path + (field.name,), child_value)
         full_repr = f"{field.name}: ({_repr})"
         if field.metadata.get(FIELD_METADATA_CACHE_VALUE, False):
             full_repr += f" = {child_value}"
         repr_l.append(full_repr)
     return "[" + ",".join(repr_l) + "]"
 
 
-def stringify_obj_type(obj: Any) -> str | None:
+def stringify_obj_type(path: tuple[str, ...], obj: Any) -> str | None:
     """
     stringify the type of obj
 
     String should somehow represent the type of obj. Doesnt have to be hashed, nor does it have
     to be the actual python type string, just something representative of the type, and won't collide
```

Comment:
> nor does it have to be the actual python type string

What? It always return a Python string.

Reply:
yes, but let's say we have an NDArray. If you do `str(type(my_ndarray))` you'll get something liek `<class 'gstaichi.lang._ndarray.ScalarNdarray'>"`. This stringify is not required to return this. Nor does it. There is no requiremetn on the string at all, except that it somehow uniquely identifies the type, relative to other types stringified by this fucntion. So "nd" is valid, so is "abc", etc.